### PR TITLE
Add Scrolling To Text Boxes

### PR DESCRIPTION
**Pull request by TheBrokenRail**
[Screencast from 2024-05-14 16-20-49.webm](https://github.com/ReMinecraftPE/mcpe/assets/17478432/1e97a58a-d193-4bc8-876c-97bae55a65fb)

### CODE REVIEW

**Review by BrentDaMage — APPROVED** [review]
Looks all good.

File: source/client/gui/components/TextInputBox.cpp [hunk 1]
```diff
@@ -217,99 +210,82 @@ void TextInputBox::onClick(int x, int y)
...
+	m_insertHead++;
+	recalculateScroll();
+}
 
-	// note: the width will increase by the same amount no matter where K is appended
-	int width = m_pFont->width(m_text + char(k));
-	if (width < m_width - 2)
+static int PADDING = 5;
+
+static std::string get_rendered_text(Font *font, int width, int scroll_pos, std::string text)
```
Comment (iProgramMC) [inline source/client/gui/components/TextInputBox.cpp]: `getRenderedText`.  Also might I suggest you make it part of the control instead?

File: source/client/gui/components/TextInputBox.cpp [hunk 2]
```diff
@@ -217,99 +210,82 @@ void TextInputBox::onClick(int x, int y)
...
+	// Insert
+	m_text.insert(m_text.begin() + m_insertHead, k);
+	m_insertHead++;
+	recalculateScroll();
+}
 
-	// note: the width will increase by the same amount no matter where K is appended
-	int width = m_pFont->width(m_text + char(k));
-	if (width < m_width - 2)
+static int PADDING = 5;
```
Comment (iProgramMC) [inline source/client/gui/components/TextInputBox.cpp]: Use `constexpr` (defined as const in C++0x mode), or `const`.

File: source/client/gui/components/TextInputBox.cpp [hunk 3]
```diff
@@ -217,99 +210,82 @@ void TextInputBox::onClick(int x, int y)
...
+	while (font->width(rendered_text) > max_width)
 	{
-		m_text.insert(m_text.begin() + m_insertHead, k);
-		m_insertHead++;
+		rendered_text.pop_back();
 	}
+	return rendered_text;
 }
 
+static char CURSOR_CHAR = '_';
```
Comment (iProgramMC) [inline source/client/gui/components/TextInputBox.cpp]: Again, `const` or `constexpr`.

File: source/client/gui/components/TextInputBox.cpp [hunk 4]
```diff
@@ -47,117 +45,111 @@ void TextInputBox::setEnabled(bool bEnabled)
...
 
-void TextInputBox::keyPressed(Minecraft* minecraft, int key)
+void TextInputBox::keyPressed(int key)
 {
 	if (!m_bFocused)
-		return;
-
-	bool bShiftPressed = minecraft->platform()->shiftPressed();
-	
-#ifndef HANDLE_CHARS_SEPARATELY
```
Comment (iProgramMC) [inline source/client/gui/components/TextInputBox.cpp:57]: `HANDLE_CHARS_SEPARATELY` is not always available.  Sure, all platforms available use it currently, but it's possible that that could change.

**Review by iProgramMC — CHANGES_REQUESTED** [review]

File: source/client/gui/components/TextInputBox.cpp (hunk 4 above)
Comment (TheBrokenRail) [reply source/client/gui/components/TextInputBox.cpp:57]: Done!

File: source/client/gui/components/TextInputBox.cpp (hunk 3 above)
Comment (TheBrokenRail) [reply source/client/gui/components/TextInputBox.cpp]: Done!

File: source/client/gui/components/TextInputBox.cpp (hunk 2 above)
Comment (TheBrokenRail) [reply source/client/gui/components/TextInputBox.cpp]: Done!

File: source/client/gui/components/TextInputBox.cpp (hunk 1 above)
Comment (TheBrokenRail) [reply source/client/gui/components/TextInputBox.cpp]: Done!

File: source/client/gui/components/TextInputBox.cpp [hunk 5]
```diff
@@ -217,99 +301,82 @@ void TextInputBox::onClick(int x, int y)
...
-		std::string substr = m_text.substr(0, m_insertHead);
-		xPos += m_pFont->width(substr);
+		int cursor_pos = m_insertHead - m_scrollPos;
+		if (cursor_pos >= 0 && cursor_pos <= int(rendered_text.length()))
+		{
+			std::string substr = rendered_text.substr(0, cursor_pos);
+			int xPos = PADDING + m_pFont->width(substr);
 
-		drawString(m_pFont, "_", m_xPos + xPos, m_yPos + textYPos + 2, 0xFFFFFF);
+			std::string str;
```
Comment (iProgramMC) [inline source/client/gui/components/TextInputBox.cpp:376]: FYI, `std::string(1, CURSOR_CHAR)`.  Just a minor nitpick though.

**Review by iProgramMC — APPROVED** [review]

**Pull request merged**